velox/functions/sparksql/String.h

```diff
@@ -1556,4 +1556,44 @@ struct Empty2NullFunction {
   }
 };
 
+/// luhn_check(input) -> boolean
+///
+/// Checks that a string of digits is valid according to the Luhn algorithm.
+template <typename T>
+struct LuhnCheckFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<bool>& result,
+      const arg_type<Varchar>& input) {
+    // Empty string is not a valid Luhn number.
+    if (input.empty()) {
+      result = false;
+      return;
+    }
+
+    int checkSum = 0;
+    bool isSecond = false;
+
+    for (auto it = input.end(); it != input.begin();) {
+      --it;
+      if (!std::isdigit(*it)) {
+        result = false;
+        return;
+      }
+
+      const int digit = *it - '0';
+      // Double the digit if it's the second digit in the sequence.
+      const int doubled = isSecond ? digit * 2 : digit;
+      // Add the two digits of the doubled number to the sum.
+      checkSum += doubled % 10 + doubled / 10;
```

Comment:
May the checkSum overflow here?

Reply:
But it is a accumulated checkSum, if the input is very long, it may overflow.

Reply:
Please refer to https://www.geeksforgeeks.org/luhn-algorithm/. If the input is very very long, it may overflow. If change to use long, just delay the risk. Could you share any suggestion?

Reply:
Make sure if overflow occurs, the behavior is same with Spark.

Reply:
Can you try the spark sql? Generate the case, velox will return an unexpected value for checkSum with this implementation, next step is also an unexpected behavior, I'm not sure if spark can returns null or throw exception here.

Reply:
If not, here is an security vulnerability in Spark

Reply:
I dig into this. In spark, if want to overflow, we need string with 238609295 characters `9`, which comes from INT_MAX divide by 9(max value of `doubled % 10 + doubled / 10`). And max length of string with characters `9` is 65534[1].  It's no possible to overflow.

[1] https://github.com/openjdk/jdk/blob/3140de411bb55604f8a9e2484798098d9d141002/src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/PoolWriter.java#L80

Reply:
The max input string length is INT_MAX,  https://stackoverflow.com/questions/816142/strings-maximum-length-in-java-calling-length-method, for each char, the checkSum may add 9, so the max value is 9 * INT_MAX, it may overflow.

Reply:
In spark, I didn't see the document describes the limitation for string type, but I cannot believe the limitation is 65534, for the json object, it is not enough. https://github.com/kamu-data/kamu-engine-spark/issues/10

Reply:
Anyway, Gluten offloads the Spark query to velox, velox StringView supports size of INT_MAX, so it might overflow in Velox, so we should check it.